subiquity/server/tests/test_runner.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+from unittest.mock import patch
+
+from subiquitycore.tests import SubiTestCase
+from subiquity.server.runner import (
+    LoggedCommandRunner,
+    DryRunCommandRunner,
+    )
+
+
+class TestLoggedCommandRunner(SubiTestCase):
+    def setUp(self):
+        self.runner = LoggedCommandRunner(ident="my-identifier")
+
+    def test_init(self):
+        self.assertEqual(self.runner.ident, "my-identifier")
+
+    def test_forge_systemd_cmd(self):
+        environ = {
+            "PATH": "/snap/subiquity/x1/bin",
+            "PYTHONPATH": "/usr/lib/python3.8/site-packages",
+            "PYTHON": "/snap/subiquity/x1/usr/bin/python3.8",
+            "TARGET_MOUNT_POINT": "/target",
+            "SNAP": "/snap/subiquity/x1",
+        }
+
+        with patch.dict(os.environ, environ):
+            cmd = self.runner._forge_systemd_cmd(
+                    ["/bin/ls", "/root"],
+                    private_mounts=True)
+
+        expected = [
+            "systemd-run",
+            "--wait", "--same-dir",
+            "--property", "SyslogIdentifier=my-identifier",
+            "--property", "PrivateMounts=yes",
+            "--setenv", "PATH=/snap/subiquity/x1/bin",
+            "--setenv", "PYTHONPATH=/usr/lib/python3.8/site-packages",
+            "--setenv", "PYTHON=/snap/subiquity/x1/usr/bin/python3.8",
+            "--setenv", "TARGET_MOUNT_POINT=/target",
+            "--setenv", "SNAP=/snap/subiquity/x1",
+            "/bin/ls", "/root",
```

Comment:
I'm not sure asserting the order of the --setenv's is a great test but well. Perhaps there should be a test that other variables are not carried over?

Reply:
Updated. I added a variable called `DUMMY` in the mocked environment and made sure it's not in the resulting command.
Regarding the order of `--setenv` commands, I agree... I tried to split the resulting command into three distinct parts so we can do something more intelligent with the `setenv`s:
* the prefix:
```python
["systemd-run", "--wait", "--same-dir",
    "--property", "SyslogIdentifier=my-identifier",
    "--property", "PrivateMounts=yes"]
```
* the suffix (which is the supplied command prefixed with `"--"`:
```python
["--", "/bin/ls", "/root"]
```
* and what's leftover: 
```python
["--setenv", "PATH=/snap/subiquity/x1/bin",
    "--setenv", "PYTHONPATH=/usr/lib/python3.8/site-packages",
    "--setenv", "PYTHON=/snap/subiquity/x1/usr/bin/python3.8",
    "--setenv", "TARGET_MOUNT_POINT=/target",
    "--setenv", "SNAP=/snap/subiquity/x1"]
```
The implementation was based on crazy list slicing and felt way too complex for unit tests ...

Reply:
OK let's leave this like this then!